nimble/controller/src/ble_ll_nrf_raal.c

```diff
@@ -0,0 +1,243 @@
+#include "syscfg/syscfg.h"
+#include "hal/hal_timer.h"
+#include "controller/ble_ll.h"
+#include "controller/ble_ll_sched.h"
+
+#if MYNEWT_VAL(BLE_LL_NRF_RAAL_ENABLE)
+
+/*
+ * Number of ticks to add at the end of scheduled item to reconfigure radio
+ * after exiting from "external" slot.
+ */
+#define NRF_RAAL_SLOT_END_MARGIN_TICKS      (2)
+
+/* Scheduler item for granted slot */
+static struct ble_ll_sched_item g_ble_ll_nrf_raal_slot_sched;
+
+static struct hal_timer g_ble_ll_nrf_raal_slot_end_tmr;
+
+static int g_ble_ll_nrf_raal_continuous;
+
+/* XXX need locking? */
+static int g_ble_ll_nrf_raal_critical;
+
+#define BLE_LL_NRF_RAAL_PENDING_SLOT_ENTER      (0x01)
+#define BLE_LL_NRF_RAAL_PENDING_SLOT_EXIT       (0x02)
+
+static int g_ble_ll_nrf_raal_pending;
+
+extern void MYNEWT_VAL(BLE_LL_NRF_RAAL_ISR_HANDLER_NAME)(void);
+extern void nrf_raal_timeslot_started(void);
+extern void nrf_raal_timeslot_ended(void);
+
+static inline bool
+ble_ll_nrf_raal_slot_granted(void)
+{
+    return ble_ll_sched_get_current_type() == BLE_LL_SCHED_TYPE_NRF_RAAL;
+}
+
+static void
+ble_ll_nrf_raal_slot_schedule(void)
+{
+    struct ble_ll_sched_item *sch;
+    int rc;
+
+    sch = &g_ble_ll_nrf_raal_slot_sched;
+
+    if (!g_ble_ll_nrf_raal_continuous || sch->enqueued) {
+        return;
+    }
+
+    sch->start_time = os_cputime_get32();
+    sch->end_time = sch->start_time +
+                    ble_ll_usecs_to_ticks_round_up(MYNEWT_VAL(BLE_LL_NRF_RAAL_SLOT_LENGTH)) +
+                    NRF_RAAL_SLOT_END_MARGIN_TICKS;
+
+    rc = ble_ll_sched_nrf_raal(&g_ble_ll_nrf_raal_slot_sched);
+    assert(rc == 0);
+}
+
+static void
+ble_ll_nrf_raal_slot_enter(void)
+{
+    if (!g_ble_ll_nrf_raal_critical) {
+        nrf_raal_timeslot_started();
+        return;
+    }
+
+    /*
+     * No need to make enter pending if there was exit pending since we are
+     * still in slot (so basically we are still in slot).
+     */
+    if (g_ble_ll_nrf_raal_pending & BLE_LL_NRF_RAAL_PENDING_SLOT_EXIT) {
+        /*
+         * XXX is this legal? it means client was not notified that it was out
+         * of slot for some time... not yet sure what this means.
+         */
+        g_ble_ll_nrf_raal_pending = 0;
+    } else {
+        g_ble_ll_nrf_raal_pending |= BLE_LL_NRF_RAAL_PENDING_SLOT_ENTER;
+    }
+}
+
+static void
+ble_ll_nrf_raal_slot_exit(void)
+{
+    if (!g_ble_ll_nrf_raal_critical) {
+        nrf_raal_timeslot_ended();
+        return;
+    }
+
+    /*
+     * No need to make exit pending if there was enter pending since we are
+     * outside slot (so basically slot was missed).
+     */
+    if (g_ble_ll_nrf_raal_pending & BLE_LL_NRF_RAAL_PENDING_SLOT_ENTER) {
+        g_ble_ll_nrf_raal_pending = 0;
+    } else {
+        g_ble_ll_nrf_raal_pending |= BLE_LL_NRF_RAAL_PENDING_SLOT_EXIT;
+    }
+}
+
+static void
+ble_ll_nrf_raal_slot_end_tmr_cb(void *arg)
+{
+    ble_ll_nrf_raal_slot_exit();
+    ble_phy_nrf_raal_slot_exit();
+    /* XXX fixme :) */
```

Comment:
Hey, can you give me some detail on what is the issue here?

Reply:
I was not sure what to do with rfclk here (i.e. if we can just disable it) so left it as fixme to figure out this later